### PR TITLE
terminfo: symlink terminfo to /etc for ncurses

### PR DESCRIPTION
**Pull request by layus**
Terminfo was bundled in the system path, but not made available via /etc/terminfo
This PR also collects all the scattered config in nixos/modules/config/terminfo.nix

This makes #19785 work for system-wide packages on NixOS.

### CODE REVIEW

File: nixos/modules/config/terminfo.nix
Comment (Mic92) [inline nixos/modules/config/terminfo.nix]: Is `TERM` updated locally but not exported child processes?

File: pkgs/development/libraries/ncurses/default.nix
Comment (Mic92) [inline pkgs/development/libraries/ncurses/default.nix]: Is this a mass rebuild?

Comment (layus) [inline pkgs/development/libraries/ncurses/default.nix]: It is, indeed. staging would be okay too.
That being said, it is a trivial change, there is no failure to expect here.

File: nixos/modules/config/terminfo.nix
Comment (0xABAB) [inline nixos/modules/config/terminfo.nix]: I am not sure whether everyone would want to have this on every machine. Adding some knob to disable this would help. Others probably also have an opinion on this.

File: pkgs/development/libraries/ncurses/default.nix
Comment (0xABAB) [inline pkgs/development/libraries/ncurses/default.nix]: You don't need to quote here.

**Review by 0xABAB — CHANGES_REQUESTED** [review]
Minor things, and if nobody cares about the sudo thing, then I also don't care.

File: nixos/modules/config/terminfo.nix
Comment (layus) [inline nixos/modules/config/terminfo.nix]: I have just gathered scattered options in the same file. This is not a new feature. Adding an option could be good, but that's not related to this PR.

File: pkgs/development/libraries/ncurses/default.nix
Comment (layus) [inline pkgs/development/libraries/ncurses/default.nix]: Well, they cannot hurt, right ? Also, this is not part of the changeset :-).

Comment (0xABAB) [inline pkgs/development/libraries/ncurses/default.nix]: Yeah, that seems to be a recurring thing. I would like people to fix things when they are glaring in my face, and preferably also before they do a pull request for new features. I.e., to go over whatever wrong things the person before you has done and to fix it.

I.e., instead of arguing over it, wouldn't it be faster to just fix it and amend a commit? The alternative would be for me to create a PR specifically to fix whatever stupid quoting people do.

Comment (Ralith) [inline pkgs/development/libraries/ncurses/default.nix]: Why isn't quoting here appropriate? Where is it guaranteed that `dev` will never contain any characters that require quoting?

File: nixos/modules/config/terminfo.nix
Comment (Mic92) [inline nixos/modules/config/terminfo.nix]: For context: the original motivation to disable this in sudo goes back to a bug in the 2000s: http://securityvulns.com/Adocument760.html

File: pkgs/development/libraries/ncurses/default.nix
Comment (4z3) [inline pkgs/development/libraries/ncurses/default.nix]: Quoting _is_ appropriate here because if expanded `$dev` needs quoting, then it will break the script. Quoting can be safely omitted when using `export` like this: `a=$b; export a`.

Comment (0xABAB) [inline pkgs/development/libraries/ncurses/default.nix]: @Ralith If you would be so kind to tell me in the kindest possible way that your assumptions are wrong, then I can reply to you in a kind way.

Comment (Ralith) [inline pkgs/development/libraries/ncurses/default.nix]: If answering the question doesn't work for you, you could try identifying the wrong assumption and explaining the error therein.

File: nixos/modules/config/terminfo.nix
Comment (layus) [inline nixos/modules/config/terminfo.nix]: Do you mean that it is now safe to export that env var for everyone ?

File: pkgs/development/libraries/ncurses/default.nix
Comment (layus) [inline pkgs/development/libraries/ncurses/default.nix]: This is taking too much attention/time.
Quoting here is correct, is resistant to future changes, and is not an issue of any kind. Plus it is not part of the modified part of this PR. Changing this will pollute git blame with no extra benefit to execution or readability.

I will not modify this PR. I guess that https://github.com/NixOS/nixpkgs/issues/26951 is a better place to discuss the general issue than here.

File: nixos/modules/config/terminfo.nix
Comment (0xABAB) [inline nixos/modules/config/terminfo.nix]: @layus No, it means that it depends on your risk appetite.

File: pkgs/development/libraries/ncurses/default.nix
Comment (0xABAB) [inline pkgs/development/libraries/ncurses/default.nix]: I already said that the quotes are redundant, which follows from the semantics of POSIX sh. The second question's answer is that there is no such guarantee, but it's irrelevant.

Comment (layus) [inline pkgs/development/libraries/ncurses/default.nix]: Because you insisted, I made several tests with different shells and indeed:

```bash
a="a B=b"
export A=$a
echo $B # =>
echo $A # => a B=b
```

I would consider myself quite knowledgeable in bash, but I have no idea why this works so.

Bash documentation has 
```
DEFINITIONS
       [...]
       word   A sequence of characters considered as a single unit by the shell.  Also known as a token.
       name   A word consisting only of alphanumeric characters and underscores, and beginning with an alphabetic  character  or  an  underscore.
              Also referred to as an identifier.
       [...]

BUILTIN COMMANDS
       [...]
       export [-fn] [name[=word]] ...
       export -p
              The supplied names are marked for automatic export to the environment of subsequently executed  commands.   If  the  -f  option  is
              given, the names refer to functions.  If no names are given, or if the -p option is supplied, a list of names of all exported vari‐
              ables is printed.  The -n option causes the export property to be removed from each name.  If a variable name is followed by =word,
              the  value  of the variable is set to word.  export returns an exit status of 0 unless an invalid option is encountered, one of the
              names is not a valid shell variable name, or -f is supplied with a name that is not a function.
```

There is no clue in there that it works as it does. The actual semantics is more like variable affectations optionally preceded by the export keyword.

Bash is a strange language, with a lot of pitfalls. As you can see from both Ralith and me, this corner case is not well known. We assumed that variable expansion occurs here like it does for other commands. The builtin implementation behave differently, but is compatible with our mental model. Why not go the safe side ? Especially when it is not clearly documented.

Comment (4z3) [inline pkgs/development/libraries/ncurses/default.nix]: IIUC, then unquoted values in `export a=$b` isn't yet part of POSIX. But bash does already support it (dash doesn't, but we care only for bash in nixpkgs, right? :))

Refs: https://github.com/koalaman/shellcheck/issues/595, http://austingroupbugs.net/view.php?id=351#c943

Comment (layus) [inline pkgs/development/libraries/ncurses/default.nix]: Thank you @4z3 for providing references. I was unable to find any.
Now I really refuse to make a change according to an undocumented implementation of an unpublished (and not freely available) standard ;-).
I also agree that $out should never contain spaces, but again, look at the length of this discussion just to agree that not quoting is correct, while everybody agrees easily that quoting is correct.

Comment (4z3) [inline pkgs/development/libraries/ncurses/default.nix]: It is available at http://pubs.opengroup.org/onlinepubs/9699919799/

File: nixos/modules/config/terminfo.nix
Comment (layus) [inline nixos/modules/config/terminfo.nix]: Well, wee need to collect more feedback then.

Comment (layus) [inline nixos/modules/config/terminfo.nix]: Or we could simply drop these lines, as root's profile should setup TERMINFO correctly.
I will do that in a separate PR, and notify the ML.

File: pkgs/development/libraries/ncurses/default.nix
Comment (0xABAB) [inline pkgs/development/libraries/ncurses/default.nix]: @layus It has been like this since the 1990s. That you don't know this, nor the others, is becoming annoying.

Comment (layus) [inline pkgs/development/libraries/ncurses/default.nix]: Well, it is always nice to learn new things, and I really enjoy getting to the bottom of an issue/disagreement. You are indeed correct that the quotes are not needed here.
I think I can also understand your irritation at seeing code that does not match your standards. I also expect contributors to fix issues they encounter around their patches.

What I do not understand is why you consider extra quoting a big issue. It is common practice to quote bash strings, even when not needed. This shows that the content is intended to be a string. It is a bit like curly braces for blocks in C. If your block is only one statement, curly braces are not needed, but nobody ever asks you to remove them. While redundant, they are correct and convey intent. 
The same applies with export: `export CFLAGS="-Wall"` does not need quotes, but if you add another argument, then you will want the quotes, as in `export CFLAGS="-Wall -O1".

I will not change this, because it does not belong to this commit, and is not needed. You are of course free to fix it in another PR, but I urge you not to do that. While you really want to get rid of useless quotes, there will be someone else out there that insists on quoting every string. In the absence of a clear consensus, this will end up in a load of commits toggling quotes, with no added value. It will be even worse if these toggle happen together with unrelated changes.
For now, there is about 600 `export` statements with quotes in nixpkgs, and the same without quotes. No consensus.

I really want to make a point here, because I did not appreciate being called on this non error. Like everyone else, I try to make my PR as good as possible and I appreciate constructive comments and technical discussions. I however find that subjective comments about code formatting make submitting PRs a pain. I do not understand why you would want to lose time even noticing that, and I am afraid that you may make the same comments to other contributors. 
Can you understand how irritating it is to get these comments when members commit code like that all the time ? Even Eelco does that sometimes (3fe93d2f75b36a2663a377826fc792b8207c3f7b, 452afd1ed68a53e5c845f9df96d707b24d40b250, 81e530a7499e3f06bec13b6ec974d0f1e3a4ab84) even though not all the time, and sometimes alternating style in the same commit (497c828611f248baa8bdf8320e841b24e528bbe4). From his coding style, I am not even sure he knows about the above POSIX rules around export.

I understand that you may be annoyed, but it should no be at the expense of random contributors that just happen not to have direct commit access. If you want this to change, please make it a general discussion on the ML.

**Comment by Mic92** [comment]
Please rebase this for staging.

**Comment by layus** [comment]
@Mic92 Rebased, and re-targeted the PR.

**Comment by Mic92** [comment]
ping @vcunat

**Comment by Mic92** [comment]
merging should be probably delayed until: https://github.com/NixOS/nixpkgs/issues/19785#issuecomment-312252923
is integrated.

**Comment by vcunat** [comment]
I'm afraid I know nothing about terminfo and almost nothing about ncurses.

**Comment by Mic92** [comment]
It was only a reminder from side, since I do not have the permission to coordinate staging builds.

**Comment by vcunat** [comment]
Ah, I see what you meant.  I didn't understand it only waited to be merged.  (I didn't read the whole thread.)

**Comment by Ralith** [comment]
Isn't https://github.com/NixOS/nixpkgs/issues/19785#issuecomment-312252923 independent of this? My understanding is that it fixes zsh's intended magic handling of the relevant environment variables to Just Work when they're updated live, whereas this PR fixes things that lack such magic.

**Comment by Mic92** [comment]
yes, after realizing that, I merged this pull request.

**Comment by edolstra** [comment]
It isn't actually clear to me why we need to set $TERMINFO_DIRS at all. Ncurses looks in its own $out/share/terminfo by default, so it seems to me that it can find terminfo files just fine without $TERMINFO_DIRS or /etc/terminfo.

Theoretically, $TERMINFO_DIRS allows people to install additional terminfo files in their own profile or in the system profile, but anybody still need that in 2017?

Regarding #19785 the real question is why zsh doesn't use $ncurses/share/terminfo. A quick strace shows zsh finding terminfo files just fine when $TERMINFO_DIRS is unset:
```
$ strace ./result/bin/zsh 2>&1 | grep terminfo
stat("/home/eelco/.terminfo", 0x21bc1c0) = -1 ENOENT (No such file or directory)
stat("/nix/store/hq2knvazvxrkf3gy49r5v51r8xq1i6qh-ncurses-6.0/share/terminfo", {st_mode=S_IFDIR|0555, st_size=4096, ...}) = 0
access("/nix/store/hq2knvazvxrkf3gy49r5v51r8xq1i6qh-ncurses-6.0/share/terminfo/x/xterm-256color", R_OK) = 0
open("/nix/store/hq2knvazvxrkf3gy49r5v51r8xq1i6qh-ncurses-6.0/share/terminfo/x/xterm-256color", O_RDONLY) = 11
% 
```

**Comment by Ralith** [comment]
> Theoretically, $TERMINFO_DIRS allows people to install additional terminfo files in their own profile or in the system profile, but anybody still need that in 2017?

Yes. Many widely used terminals come with their own terminfo, and will not gracefully degrade in its absence. I have personally been repeatedly frustrated by this issue when using urxvt and termite, for example. See discussion in https://github.com/NixOS/nixpkgs/issues/19785.

**Pull request merged**
